File: Android/src/jp/co/riso/smartdeviceapp/view/fragment/PrintSettingsFragment.java
/*
 * Copyright (c) 2014 RISO, Inc. All rights reserved.
 *
 * PrintSettingsFragment.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */

package jp.co.riso.smartdeviceapp.view.fragment;

import java.util.Date;
import java.util.Locale;

import jp.co.riso.android.dialog.DialogUtils;
import jp.co.riso.android.dialog.InfoDialogFragment;
import jp.co.riso.android.dialog.WaitingDialogFragment;
import jp.co.riso.android.dialog.WaitingDialogFragment.WaitingDialogListener;
import jp.co.riso.android.os.pauseablehandler.PauseableHandler;
import jp.co.riso.android.os.pauseablehandler.PauseableHandlerCallback;
import jp.co.riso.android.util.AppUtils;
import jp.co.riso.android.util.NetUtils;
import jp.co.riso.smartprint.R;
import jp.co.riso.smartdeviceapp.SmartDeviceApp;
import jp.co.riso.smartdeviceapp.common.DirectPrintManager;
import jp.co.riso.smartdeviceapp.common.DirectPrintManager.DirectPrintCallback;
import jp.co.riso.smartdeviceapp.controller.jobs.PrintJobManager;
import jp.co.riso.smartdeviceapp.controller.pdf.PDFFileManager;
import jp.co.riso.smartdeviceapp.controller.printer.PrinterManager;
import jp.co.riso.smartdeviceapp.model.PrintJob.JobResult;
import jp.co.riso.smartdeviceapp.model.Printer;
import jp.co.riso.smartdeviceapp.model.Printer.PortSetting;
import jp.co.riso.smartdeviceapp.model.printsettings.PrintSettings;
import jp.co.riso.smartdeviceapp.view.base.BaseFragment;
import jp.co.riso.smartdeviceapp.view.printsettings.PrintSettingsView;
import android.os.Bundle;
import android.os.Message;
import android.view.View;
import android.widget.TextView;

public class PrintSettingsFragment extends BaseFragment implements PrintSettingsView.PrintSettingsViewInterface, PauseableHandlerCallback, DirectPrintCallback, WaitingDialogListener {
    
    private static final String TAG_WAITING_DIALOG = "dialog_printing";
    private static final String TAG_MESSAGE_DIALOG = "dialog_message";
    
    private DirectPrintManager mDirectPrintManager = null;
    
    private static final int MSG_PRINT = 0;
    
    private boolean mFragmentForPrinting = false;
    
    private int mPrinterId = PrinterManager.EMPTY_ID;
    private PrintSettings mPrintSettings;
    private PrintSettingsView mPrintSettingsView;
    private Bundle mPrintSettingsBundle = null;
    
    private String mPdfPath;
    private PauseableHandler mPauseableHandler;
    private WaitingDialogFragment mWaitingDialog;
    private String mPrintMsg = "";
    
    /** {@inheritDoc} */
    @Override
    public int getViewLayout() {
        return R.layout.fragment_printsettings;
    }
    
    /** {@inheritDoc} */
    @Override
    public void initializeFragment(Bundle savedInstanceState) {
        setRetainInstance(true);
        
        if (mPrintSettings == null) {
            mPrintSettings = new PrintSettings();
        }
        if (mPauseableHandler == null) {
            mPauseableHandler = new PauseableHandler(this);
        }
        
        mPrintMsg = getResources().getString(R.string.ids_info_msg_printing);
    }
    
    /** {@inheritDoc} */
    @Override
    public void initializeView(View view, Bundle savedInstanceState) {
        mPrintSettingsView = (PrintSettingsView) view.findViewById(R.id.rootView);
        
        mPrintSettingsView.setValueChangedListener(this);
        
        mPrintSettingsView.setInitialValues(mPrinterId, mPrintSettings);
        mPrintSettingsView.setShowPrintControls(mFragmentForPrinting);
        
        TextView textView = (TextView) view.findViewById(R.id.titleTextView);
        textView.setText(R.string.ids_lbl_print_settings);
        
        if (!mFragmentForPrinting) {
            textView.setText(R.string.ids_lbl_default_print_settings);
        }
        
        if (mPrintSettingsBundle != null) {
            mPrintSettingsView.restoreState(mPrintSettingsBundle);
            mPrintSettingsBundle = null;
        }
    }
    
    /** {@inheritDoc} */
    @Override
    public void initializeCustomActionBar(View view, Bundle savedInstanceState) {
    }
    
    /** {@inheritDoc} */
    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        
        if (mPrintSettingsView != null) {
            mPrintSettingsBundle = new Bundle();
            mPrintSettingsView.saveState(mPrintSettingsBundle);
        }
    }
    
    /** {@inheritDoc} */
    @Override
    public void onPause() {
        super.onPause();
        
        PrinterManager.getInstance(SmartDeviceApp.getAppContext()).cancelUpdateStatusThread();
        mPauseableHandler.pause();
    }
    
    /** {@inheritDoc} */
    @Override
    public void onResume() {
        super.onResume();
        mPauseableHandler.resume();
    }
    
    // ================================================================================
    // Public functions
    // ================================================================================
    
    /**
     * Set fragment for printing
     * 
     * @param fragmentForPrinting
     */
    public void setFragmentForPrinting(boolean fragmentForPrinting) {
        mFragmentForPrinting = fragmentForPrinting;
    }
    
    /**
     * Set printer ID
     * 
     * @param printerId
     */
    public void setPrinterId(int printerId) {
        mPrinterId = printerId;
    }
    
    /**
     * Set print settings
     * 
     * @param printSettings
     */
    public void setPrintSettings(PrintSettings printSettings) {
        mPrintSettings = new PrintSettings(printSettings);
    }
    
    /**
     * This method sets the value of mPdfPath
     * 
     * @param pdfPath
     *            the PDF sandbox path
     */
    public void setPdfPath(String pdfPath) {
        mPdfPath = pdfPath;
    }
    
    // ================================================================================
    // INTERFACE - ValueChangedListener
    // ================================================================================
    
    /** {@inheritDoc} */
    @Override
    public void onPrinterIdSelectedChanged(int printerId) {
        setPrinterId(printerId);
        
        if (getTargetFragment() instanceof PrintPreviewFragment) {
            PrintPreviewFragment fragment = (PrintPreviewFragment) getTargetFragment();
            fragment.setPrintId(printerId);
        }
    }
    
    /** {@inheritDoc} */
    @Override
    public void onPrintSettingsValueChanged(PrintSettings printSettings) {
        setPrintSettings(printSettings);
        
        if (!mFragmentForPrinting) {
            printSettings.savePrintSettingToDB(mPrinterId);
        }
        
        if (getTargetFragment() instanceof PrintPreviewFragment) {
            PrintPreviewFragment fragment = (PrintPreviewFragment) getTargetFragment();
            fragment.setPrintSettings(printSettings);
        }
    }
    
    /**
     * This method is triggered when the print button is pressed and displays the waiting dialog.
     */
    @Override
    public void onPrint(Printer printer, PrintSettings printSettings) {
        // do not print if mPdfPath is not set
        if (mPdfPath != null && mPdfPath.isEmpty()) {
            return;
        }
        
        if (printer == null || printSettings == null) {
            String strMsg = getString(R.string.ids_err_msg_no_selected_printer);
            String btnMsg = getString(R.string.ids_lbl_ok);
            InfoDialogFragment fragment = InfoDialogFragment.newInstance(strMsg, btnMsg);
            DialogUtils.displayDialog(getActivity(), TAG_MESSAGE_DIALOG, fragment);
            return;
        }
        
        if (!NetUtils.isNetworkAvailable(getActivity())) {
            String strMsg = getString(R.string.ids_err_msg_network_error);
            String btnMsg = getString(R.string.ids_lbl_ok);
            InfoDialogFragment fragment = InfoDialogFragment.newInstance(strMsg, btnMsg);
            DialogUtils.displayDialog(getActivity(), TAG_MESSAGE_DIALOG, fragment);
            return;
        }
        
        String btnMsg = getResources().getString(R.string.ids_lbl_cancel);
        mWaitingDialog = WaitingDialogFragment.newInstance(null, mPrintMsg, true, btnMsg);
        mWaitingDialog.setTargetFragment(this, 0);
        DialogUtils.displayDialog(getActivity(), TAG_WAITING_DIALOG, mWaitingDialog);
        
        String jobname = PDFFileManager.getSandboxPDFName(SmartDeviceApp.getAppContext());
        
        mDirectPrintManager = new DirectPrintManager();
        mDirectPrintManager.setCallback(this);
        
        
        String userName = AppUtils.getOwnerName();

        if (printer.getPortSetting() == PortSetting.LPR) {
            mDirectPrintManager.executeLPRPrint(userName, jobname, mPdfPath, printSettings.formattedString(), printer.getIpAddress());
        } else {
            mDirectPrintManager.executeRAWPrint(userName, jobname, mPdfPath, printSettings.formattedString(), printer.getIpAddress());
        }
    }
    
    // ================================================================================
    // INTERFACE - PauseableHandlerCallback
    // ================================================================================
    
    /** {@inheritDoc} */
    @Override
    public boolean storeMessage(Message message) {
        return message.what == MSG_PRINT;
    }
    
    /** {@inheritDoc} */
    @Override
    public void processMessage(Message message) {
        switch (message.what) {
            case MSG_PRINT:
                DialogUtils.dismissDialog(getActivity(), TAG_WAITING_DIALOG);
                
                PrintJobManager pm = PrintJobManager.getInstance(SmartDeviceApp.getAppContext());
                String filename = PDFFileManager.getSandboxPDFName(SmartDeviceApp.getAppContext());
                
                if (message.arg1 == DirectPrintManager.PRINT_STATUS_SENT) {
                    pm.createPrintJob(mPrinterId, filename, new Date(), JobResult.SUCCESSFUL);
                    ((PrintPreviewFragment) getFragmentManager().findFragmentById(R.id.mainLayout)).clearIconStates();
                    ((HomeFragment) getFragmentManager().findFragmentById(R.id.leftLayout)).goToJobsFragment();
                    
                    // Show dialog
                    String strMsg = getString(R.string.ids_info_msg_print_job_successful);
                    String btnMsg = getString(R.string.ids_lbl_ok);
                    InfoDialogFragment fragment = InfoDialogFragment.newInstance(strMsg, btnMsg);
                    DialogUtils.displayDialog(getActivity(), TAG_MESSAGE_DIALOG, fragment);
                }
                else {
                    pm.createPrintJob(mPrinterId, filename, new Date(), JobResult.ERROR);
                    
                    // Show dialog
                    String strMsg = getString(R.string.ids_info_msg_print_job_failed);
                    String btnMsg = getString(R.string.ids_lbl_ok);
                    InfoDialogFragment fragment = InfoDialogFragment.newInstance(strMsg, btnMsg);
                    DialogUtils.displayDialog(getActivity(), TAG_MESSAGE_DIALOG, fragment);
                }
                break;
        }
    }
    
    // ================================================================================
    // INTERFACE - DirectPrintCallback
    // ================================================================================
    
    /** {@inheritDoc} */
    @Override
    public void onNotifyProgress(DirectPrintManager manager, int status, float progress) {
        if (NetUtils.isNetworkAvailable(SmartDeviceApp.getAppContext())) {
            switch (status) {
                case DirectPrintManager.PRINT_STATUS_ERROR_CONNECTING:
                case DirectPrintManager.PRINT_STATUS_ERROR_SENDING:
                case DirectPrintManager.PRINT_STATUS_ERROR_FILE:
                case DirectPrintManager.PRINT_STATUS_ERROR:
                case DirectPrintManager.PRINT_STATUS_SENT:
                    Message newMessage = Message.obtain(mPauseableHandler, MSG_PRINT);
                    newMessage.arg1 = status;
                    mPauseableHandler.sendMessage(newMessage);
                    break;
                case DirectPrintManager.PRINT_STATUS_SENDING:
                    if (mWaitingDialog != null) {
                        String msg = String.format(Locale.getDefault(), "%s %.2f%%", mPrintMsg, progress);
                        mWaitingDialog.setMessage(msg);
                    }
                    break;
                case DirectPrintManager.PRINT_STATUS_STARTED:
                case DirectPrintManager.PRINT_STATUS_CONNECTING:
                case DirectPrintManager.PRINT_STATUS_CONNECTED:
                    break;
            }
        } else {
            // cancel Direct Print but considered as failed job
            onCancel();
            Message newMessage = Message.obtain(mPauseableHandler, MSG_PRINT);
            mPauseableHandler.sendMessage(newMessage);
        }
    }
    
    // ================================================================================
    // INTERFACE - WaitingDialogListener
    // ================================================================================
    
    /** {@inheritDoc} */
    @Override
    public void onCancel() {
        if (mDirectPrintManager != null) {
            mDirectPrintManager.sendCancelCommand();
            mDirectPrintManager = null;
        }
    }
}

File: Android/app/src/main/java/jp/co/riso/smartdeviceapp/view/fragment/PrintSettingsFragment.kt
/*
 * Copyright (c) 2022 RISO, Inc. All rights reserved.
 *
 * PrintSettingsFragment.kt
 * SmartDeviceApp
 * Created by: a-LINK Group
 */
package jp.co.riso.smartdeviceapp.view.fragment

import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.os.Message
import android.view.View
import android.widget.TextView
import androidx.core.os.bundleOf
import jp.co.riso.android.dialog.DialogUtils
import jp.co.riso.android.dialog.DialogUtils.dismissDialog
import jp.co.riso.android.dialog.DialogUtils.displayDialog
import jp.co.riso.android.dialog.InfoDialogFragment
import jp.co.riso.android.dialog.InfoDialogFragment.Companion.newInstance
import jp.co.riso.android.dialog.WaitingDialogFragment
import jp.co.riso.android.dialog.WaitingDialogFragment.Companion.newInstance
import jp.co.riso.android.dialog.WaitingDialogFragment.WaitingDialogListener
import jp.co.riso.android.os.pauseablehandler.PauseableHandler
import jp.co.riso.android.os.pauseablehandler.PauseableHandlerCallback
import jp.co.riso.android.util.AppUtils
import jp.co.riso.android.util.NetUtils.isWifiAvailable
import jp.co.riso.smartdeviceapp.SmartDeviceApp
import jp.co.riso.smartdeviceapp.common.DirectPrintManager
import jp.co.riso.smartdeviceapp.common.DirectPrintManager.DirectPrintCallback
import jp.co.riso.smartdeviceapp.controller.jobs.PrintJobManager
import jp.co.riso.smartdeviceapp.controller.pdf.PDFFileManager.Companion.getSandboxPDFName
import jp.co.riso.smartdeviceapp.controller.printer.PrinterManager
import jp.co.riso.smartdeviceapp.model.PrintJob.JobResult
import jp.co.riso.smartdeviceapp.model.Printer
import jp.co.riso.smartdeviceapp.model.Printer.PortSetting
import jp.co.riso.smartdeviceapp.model.printsettings.PrintSettings
import jp.co.riso.smartdeviceapp.view.base.BaseFragment
import jp.co.riso.smartdeviceapp.view.printsettings.PrintSettingsView
import jp.co.riso.smartdeviceapp.view.printsettings.PrintSettingsView.PrintSettingsViewInterface
import jp.co.riso.smartprint.R
import java.util.*
import kotlin.math.min

/**
 * @class PrintSettingsFragment
 *
 * @brief Fragment which contains the Print Settings Screen
 */
class PrintSettingsFragment : BaseFragment(), PrintSettingsViewInterface, PauseableHandlerCallback,
    DirectPrintCallback, WaitingDialogListener {
    private var _directPrintManager: DirectPrintManager? = null
    private var _fragmentForPrinting = false
    private var _printerId = PrinterManager.EMPTY_ID
    private var _printSettings: PrintSettings? = null
    private var _printSettingsView: PrintSettingsView? = null
    private var _printSettingsBundle: Bundle? = null
    private var _pdfPath: String? = null
    private var _pdfIsLandscape = false
    private var _pauseableHandler: PauseableHandler? = null
    private var _waitingDialog: WaitingDialogFragment? = null
    private var _printMsg = ""

    override val viewLayout: Int
        get() = R.layout.fragment_printsettings

    override fun initializeFragment(savedInstanceState: Bundle?) {
        retainInstance = true
        if (_printSettings == null) {
            _printSettings = PrintSettings()
        }
        if (_pauseableHandler == null) {
            _pauseableHandler = PauseableHandler(Looper.myLooper(), this)
        }
        _printMsg = resources.getString(R.string.ids_info_msg_printing)
    }

    override fun initializeView(view: View, savedInstanceState: Bundle?) {
        _printSettingsView = view.findViewById(R.id.rootView)
        _printSettingsView!!.setValueChangedListener(this)
        _printSettingsView!!.setInitialValues(_printerId, _printSettings!!)
        _printSettingsView!!.setShowPrintControls(_fragmentForPrinting)
        val textView = view.findViewById<TextView>(R.id.titleTextView)
        textView!!.setText(R.string.ids_lbl_print_settings)
        if (!_fragmentForPrinting) {
            textView.setText(R.string.ids_lbl_default_print_settings)
        }
        if (_printSettingsBundle != null) {
            _printSettingsView!!.restoreState(_printSettingsBundle!!)
            _printSettingsBundle = null
        }
    }

    override fun initializeCustomActionBar(view: View, savedInstanceState: Bundle?) {}

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (_printSettingsView != null) {
            _printSettingsBundle = Bundle()
            _printSettingsView!!.saveState(_printSettingsBundle!!)
        }
    }

    override fun onPause() {
        super.onPause()
        PrinterManager.getInstance(SmartDeviceApp.appContext!!)!!.cancelUpdateStatusThread()
        _pauseableHandler!!.pause()
    }

    override fun onResume() {
        super.onResume()
        _pauseableHandler!!.resume()

        //update strings in case locale has changed         
        _printMsg = getString(R.string.ids_info_msg_printing)
        if (_waitingDialog != null) {
            _waitingDialog!!.setButtonText(getString(R.string.ids_lbl_cancel))
        }
    }
    // ================================================================================
    // Public functions
    // ================================================================================
    /**
     * @brief Sets fragment for printing
     *
     * @param fragmentForPrinting Whether the fragment is for printing or for default print settings
     */
    fun setFragmentForPrinting(fragmentForPrinting: Boolean) {
        _fragmentForPrinting = fragmentForPrinting
    }

    /**
     * @brief Sets the printer ID of the view
     *
     * @param printerId Printer ID
     */
    fun setPrinterId(printerId: Int) {
        _printerId = printerId
    }

    /**
     * @brief Creates a copy of the print settings to be applied as print settings values
     *
     * @param printSettings Print settings to be applied
     */
    fun setPrintSettings(printSettings: PrintSettings?) {
        _printSettings = PrintSettings(printSettings!!)
    }

    /**
     * @brief This method sets the value of mPdfPath
     *
     * @param pdfPath The path of the PDF to be printed
     */
    fun setPdfPath(pdfPath: String?) {
        _pdfPath = pdfPath
    }

    /**
     * @brief This method sets whether the PDF is landscape or not
     *
     * @note Will be passed in the PJL
     *
     * @param pdfIsLandscape Whether PDF is landscape or not
     */
    fun setPDFisLandscape(pdfIsLandscape: Boolean) {
        _pdfIsLandscape = pdfIsLandscape
    }

    // ================================================================================
    // INTERFACE - ValueChangedListener
    // ================================================================================
    override fun onPrinterIdSelectedChanged(printerId: Int) {
        setPrinterId(printerId)

        // set printerId to PrintPreviewFragment
        requireActivity().supportFragmentManager.setFragmentResult(
            PrintPreviewFragment.TAG_RESULT_PRINT_SETTINGS,
            bundleOf(
                RESULT_KEY to RESULT_PRINTER_ID,
                RESULT_PRINTER_ID to printerId)
        )
    }

    override fun onPrintSettingsValueChanged(printSettings: PrintSettings?) {
        setPrintSettings(printSettings)
        if (!_fragmentForPrinting) {
            printSettings!!.savePrintSettingToDB(_printerId)
        }

        // set printSettings to PrintPreviewFragment
        requireActivity().supportFragmentManager.setFragmentResult(
            PrintPreviewFragment.TAG_RESULT_PRINT_SETTINGS,
            bundleOf(
                RESULT_KEY to RESULT_PRINTER_SETTINGS,
                RESULT_PRINTER_SETTINGS to printSettings)
        )
    }

    override fun onPrint(printer: Printer?, printSettings: PrintSettings?) {
        // do not print if mPdfPath is not set
        if (_pdfPath != null && _pdfPath!!.isEmpty()) {
            return
        }
        if (printer == null || printSettings == null) {
            val strMsg = getString(R.string.ids_err_msg_no_selected_printer)
            val btnMsg = getString(R.string.ids_lbl_ok)
            val fragment = newInstance(strMsg, btnMsg)
            displayDialog(requireActivity(), TAG_MESSAGE_DIALOG, fragment)
            return
        }
        if (!isWifiAvailable) {
            val strMsg = getString(R.string.ids_err_msg_network_error)
            val btnMsg = getString(R.string.ids_lbl_ok)
            val fragment = newInstance(strMsg, btnMsg)
            displayDialog(requireActivity(), TAG_MESSAGE_DIALOG, fragment)
            return
        }
        val btnMsg: String
        val jobName = getSandboxPDFName(SmartDeviceApp.appContext)
        _directPrintManager = DirectPrintManager()
        _directPrintManager!!.setCallback(this)
        val appName = resources.getString(R.string.ids_app_name)
        val appVersion = AppUtils.getApplicationVersion(SmartDeviceApp.appContext)
        val userName = AppUtils.ownerName
        // Ver.2.0.4.2 Start
        val hostName = Build.MODEL
        // Ver.2.0.4.2 End
        val ret: Boolean
        val formattedString = printSettings.formattedString(_pdfIsLandscape)
        ret = if (printer.portSetting == PortSetting.LPR) {
            // Ver.2.0.4.2 Start
            //ret = mDirectPrintManager.executeLPRPrint(printer.getName(), appName, appVersion, userName, jobName, mPdfPath, formattedString, printer.getIpAddress());
            _directPrintManager!!.executeLPRPrint(
                printer.name,
                appName,
                appVersion,
                userName,
                jobName,
                _pdfPath,
                formattedString,
                printer.ipAddress,
                hostName
            )
            // Ver.2.0.4.2 End
        } else {
            // Ver.2.0.4.2 Start
            //ret = mDirectPrintManager.executeRAWPrint(printer.getName(), appName, appVersion, userName, jobName, mPdfPath, formattedString, printer.getIpAddress());
            _directPrintManager!!.executeRAWPrint(
                printer.name,
                appName,
                appVersion,
                userName,
                jobName,
                _pdfPath,
                formattedString,
                printer.ipAddress,
                hostName
            )
            // Ver.2.0.4.2 End
        }
        if (ret) {
            btnMsg = getString(R.string.ids_lbl_cancel)
            _waitingDialog = newInstance(null, _printMsg, true, btnMsg, TAG_MESSAGE_DIALOG)
            setResultListenerWaitingDialog(
                requireActivity().supportFragmentManager,
                this,
                PrintPreviewFragment.TAG_WAITING_DIALOG
            )
            displayDialog(requireActivity(), TAG_WAITING_DIALOG, _waitingDialog!!)
        } else {
            val strMsg = getString(R.string.ids_info_msg_print_job_failed)
            btnMsg = getString(R.string.ids_lbl_ok)
            val fragment = newInstance(strMsg, btnMsg)
            displayDialog(requireActivity(), TAG_MESSAGE_DIALOG, fragment)
        }
    }

    // ================================================================================
    // INTERFACE - PauseableHandlerCallback
    // ================================================================================
    override fun storeMessage(message: Message?): Boolean {
        return message!!.what == MSG_PRINT
    }

    override fun processMessage(message: Message?) {
        when (message!!.what) {
            MSG_PRINT -> {
                dismissDialog(requireActivity(), TAG_WAITING_DIALOG)
                val pm = PrintJobManager.getInstance(SmartDeviceApp.appContext!!)
                val filename = getSandboxPDFName(SmartDeviceApp.appContext)
                val fragment: InfoDialogFragment
                val strMsg: String = if (message.arg1 == DirectPrintManager.PRINT_STATUS_SENT) {
                    pm!!.createPrintJob(_printerId, filename, Date(), JobResult.SUCCESSFUL)
                    getString(R.string.ids_info_msg_print_job_successful)
                } else {
                    pm!!.createPrintJob(_printerId, filename, Date(), JobResult.ERROR)
                    getString(R.string.ids_info_msg_print_job_failed)
                }
                val btnMsg: String = getString(R.string.ids_lbl_ok)
                // Show dialog
                fragment = newInstance(strMsg, btnMsg)
                displayDialog(requireActivity(), TAG_MESSAGE_DIALOG, fragment)
            }
        }
    }

    // ================================================================================
    // INTERFACE - DirectPrintCallback
    // ================================================================================
    override fun onNotifyProgress(manager: DirectPrintManager?, status: Int, progress: Float) {
        if (isWifiAvailable) {
            when (status) {
                DirectPrintManager.PRINT_STATUS_ERROR_CONNECTING, DirectPrintManager.PRINT_STATUS_ERROR_SENDING, DirectPrintManager.PRINT_STATUS_ERROR_FILE, DirectPrintManager.PRINT_STATUS_ERROR, DirectPrintManager.PRINT_STATUS_SENT -> {
                    val timerTask: TimerTask = object : TimerTask() {
                        override fun run() {
                            val newMessage = Message.obtain(_pauseableHandler, MSG_PRINT)
                            newMessage.arg1 = status
                            _pauseableHandler!!.sendMessage(newMessage)
                        }
                    }
                    val timer = Timer()
                    timer.schedule(timerTask, PRINT_JOB_SENT_PROGRESS_DIALOG_DELAY.toLong())
                }
                DirectPrintManager.PRINT_STATUS_SENDING -> if (_waitingDialog != null) {
                    val msg = String.format(
                        Locale.getDefault(), "%s %.2f%%", _printMsg, min(
                            progress, 100.0f
                        )
                    )
                    _waitingDialog!!.setMessage(msg)
                }
                DirectPrintManager.PRINT_STATUS_STARTED, DirectPrintManager.PRINT_STATUS_CONNECTING, DirectPrintManager.PRINT_STATUS_CONNECTED, DirectPrintManager.PRINT_STATUS_JOB_NUM_UPDATE -> {}
            }
        } else {
            // cancel Direct Print but considered as failed job
            onCancel()
            val newMessage = Message.obtain(_pauseableHandler, MSG_PRINT)
            _pauseableHandler!!.sendMessage(newMessage)
        }
    }

    // ================================================================================
    // INTERFACE - WaitingDialogListener
    // ================================================================================
    override fun onCancel() {
        if (_directPrintManager != null) {
            _directPrintManager!!.sendCancelCommand()
            _directPrintManager = null
        }
    }

    companion object {
        private const val TAG_WAITING_DIALOG = "dialog_printing"
        private const val TAG_MESSAGE_DIALOG = "dialog_message"
        private const val PRINT_JOB_SENT_PROGRESS_DIALOG_DELAY =
            50 // To allow user to see 100% progress percentage, enforce a 50ms delay before closing the progress dialog after a successful print job
        private const val MSG_PRINT = 0

        const val RESULT_KEY = "result"
        const val RESULT_PRINTER_ID = "printer_id"
        const val RESULT_PRINTER_SETTINGS = "printer_settings"
    }
}